oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryImpl.java

```diff
@@ -1102,13 +1113,16 @@ private SelectorExecutionPlan getBestSelectorExecutionPlan(
             IndexPlan indexPlan = null;
             if (index instanceof AdvancedQueryIndex) {
                 AdvancedQueryIndex advIndex = (AdvancedQueryIndex) index;
-                long maxEntryCount = limit;
-                if (offset > 0) {
-                    if (offset + limit < 0) {
+
+                long localLimit = limit.orElse(Long.MAX_VALUE);
+                long localOffset = offset.orElse(0L);
+                long maxEntryCount = localLimit;
+                if (localOffset > 0) {
+                    if (localOffset + localLimit < 0) {
```

Comment:
This is not really related to your patch...

Yes it works, Long.MAX_VALUE + Long.MAX_VALUE == -2.
We could also use QueryImpl.saturatedAdd, which is slower, as it's using BigInteger. 
I wonder, should we move the the logic outside of the loop?

Reply:
It's a good point though, updated